File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第10章修建公路.py
# 修建公路1  lanqiaoOJ  1124

from heapq import *
def prim():
    ans, cnt = 0, 0
    dis = [float('inf') for _ in range(n + 1)]
    dis[1] = 0
    q = []
    vis = [False for _ in range(n + 1)]   # =ture: 表示点i已经在MST中
    heappush(q, (0, 1))  #从s点开始处理队列
    while q and cnt < n:
        w, u = heappop(q)    #pop出距集合最近的点u
        if not vis[u]:
            vis[u] = True
            ans += w            
            cnt += 1
            for v,w in e[u]:      #遍历点u的邻居v，边长为w   while i:      #检查点u的所有邻居
                if dis[v] > w:
                    dis[v] = w
                    heappush(q, [dis[v], v])  #扩展新的邻居，放进优先队列
    if cnt != n:    print('-1')
    else:           print(ans)        
n, m = map(int,input().split())
e = [[] for i in range(n + 1)]
for i in range(m):
    u, v, w = map(int,input().split())
    e[u].append((v,w))   #u的邻居是v，边长w         
    e[v].append((u,w))   #双向边
prim()

File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第10章修建公路.java
//修建公路1  lanqiaoOJ 1124 

import java.io.*;
import java.util.*;
public class Main {
	static int N,M;
	static List<int[]>[] edges=new ArrayList[100001];
	static PriorityQueue<int[]> pq=new PriorityQueue<>(new Comparator<int[]>() {
		public int compare(int[] o1, int[] o2) {return o1[1]-o2[1];	}
	});
	static int[] dis;
	static int[] vis;
	static long ans;
	static int MAX=Integer.MAX_VALUE;
	static BufferedReader br=new BufferedReader(new InputStreamReader(System.in));
	static StreamTokenizer st=new StreamTokenizer(br);
	static int nextInt() throws IOException {
		st.nextToken();
		return (int)st.nval;
	}
	public static void main(String[] args) throws IOException {
			N=nextInt();
			M=nextInt();
			dis=new int[N+1];
			vis=new int[N+1];
			for (int j = 1; j <= N; j++) {
				if(edges[j]==null) {
					edges[j]=new ArrayList<>();
				}else {
					edges[j].clear();
				}
			}
			int a,b,c=0;
			for (int j = 1; j <= M; j++) {
				a=nextInt();
				b=nextInt();
				c=nextInt();
				edges[a].add(new int[] {b,c});
				edges[b].add(new int[] {a,c});
			}

			Arrays.fill(dis, MAX);
			prim();
			if(ans==-1) System.out.println("-1");
			else System.out.println(ans);
			Arrays.fill(vis,0);
			pq.clear();		
	}
	private static void prim() {
		int count = 0;
		ans = 0;
		pq.add(new int[] {1,0});
		dis[1]=0;
		while(!pq.isEmpty()) {
			int[] point=pq.poll();
			if (vis[point[0]] == 1) continue;
			vis[point[0]] = 1;
			if (point[1] != 0) {
				ans += point[1];
				count++;
			}
			for (int i = 0; i < edges[point[0]].size(); i++) {
				int[] next=edges[point[0]].get(i);
				if(next[1]<dis[next[0]]) {
					dis[next[0]]=next[1];
					pq.add(new int[] {next[0],dis[next[0]]});
				}
				
			}
		}
		if (count != N - 1) ans = -1;		
	}

}